misc/multihash/src/lib.rs

```diff
@@ -57,24 +57,48 @@ macro_rules! match_encoder {
 /// ```
 ///
 pub fn encode(hash: Hash, input: &[u8]) -> Result<Multihash, EncodeError> {
-    let (offset, mut output) = encode_hash(hash);
-    match_encoder!(hash for (input, &mut output[offset ..]) {
-        SHA1 => sha1::Sha1,
-        SHA2256 => sha2::Sha256,
-        SHA2512 => sha2::Sha512,
-        SHA3224 => sha3::Sha3_224,
-        SHA3256 => sha3::Sha3_256,
-        SHA3384 => sha3::Sha3_384,
-        SHA3512 => sha3::Sha3_512,
-        Keccak224 => sha3::Keccak224,
-        Keccak256 => sha3::Keccak256,
-        Keccak384 => sha3::Keccak384,
-        Keccak512 => sha3::Keccak512,
-        Blake2b512 => blake2::Blake2b,
-        Blake2s256 => blake2::Blake2s,
-    });
-
-    Ok(Multihash { bytes: output.freeze() })
+    // Custom length encoding for the identity multihash
+    if let Hash::Identity = hash {
+        if input.len() as u32 >= std::u32::MAX {
```

Comment:
```suggestion
        if u64::from(std::u32::MAX) < as_u64(input.len()) {
```

where `as_u64` is a little helper like this:

```rust
#[cfg(any(target_pointer_width = "32", target_pointer_width = "64"))]
fn as_u64(a: usize) -> u64 {
    a as u64
}
```

The alternative is using `TryFrom`, but it does not look pleasant.